### PR TITLE
locks to avoid concurrent token refreshes

### PR DESCRIPTION
**Pull request by eschultink**
### Features
 - parameter store based locks, use for oauth refresh token case

### Change implications

 - dependencies added/changed? **no**

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204491501062324
  - https://app.asana.com/0/0/1204295763322563

### CODE REVIEW

File: java/impl/aws/src/main/java/co/worklytics/psoxy/aws/ParameterStoreConfigService.java [hunk 1]
```diff
@@ -134,4 +141,54 @@ String parameterName(ConfigProperty property) {
             return this.namespace + property.name();
         }
     }
+
+    String lockParameterName(String lockId) {
+        return this.namespace + "lock_" + lockId;
+    }
+
+    // make configurable? YAGNI, 2m should be plenty
+    final int LOCK_TTL_SECONDS = 120;
```
Comment (jlorper) [inline java/impl/aws/src/main/java/co/worklytics/psoxy/aws/ParameterStoreConfigService.java:150]: I'd pass it as parameter

File: java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java [hunk 2]
```diff
@@ -205,9 +218,25 @@ public AccessToken refreshAccessToken() throws IOException {
             }
 
             if (shouldRefresh(this.currentToken, clock.instant())) {
+                String lockId = "oauth_refresh_token";
```
Comment (jlorper) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java:225]: extract as constant, or `this.getClass().getSimpleName()`

File: java/core/src/main/java/co/worklytics/psoxy/gateway/LockService.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package co.worklytics.psoxy.gateway;
+
+public interface LockService {
+
+    boolean acquire(String lockId);
```
Comment (jlorper) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/LockService.java:5]:
```suggestion
    boolean acquire(String lockId, long timeoutMs);
```
I'd pass it as param

File: java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java [hunk 4]
```diff
@@ -205,9 +218,25 @@ public AccessToken refreshAccessToken() throws IOException {
...
+                boolean acquired = lockService.acquire(lockId);
+
+                if (!acquired) {
+                    try {
+                        Thread.sleep(attempt * 2_000L);
```
Comment (jlorper) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java:230]: guava Uninterruptibles, we shouldn't break execution if that happens
```suggestion
Uninterruptibles.sleepUninterruptibly(attempt * 2_000L, TimeUnit.MILLISECONDS);
```

File: java/impl/aws/src/main/java/co/worklytics/psoxy/aws/ParameterStoreConfigService.java [hunk 5]
```diff
@@ -134,4 +141,54 @@ String parameterName(ConfigProperty property) {
...
+
+    @Override
+    public boolean acquire(@NonNull String lockId) {
+        final String lockParameterName = lockParameterName(lockId);
+        try {
+            client.putParameter(PutParameterRequest.builder()
+                .name(lockParameterName)
+                .type(ParameterType.STRING)
+                .value("locked")
```
Comment (jlorper) [inline java/impl/aws/src/main/java/co/worklytics/psoxy/aws/ParameterStoreConfigService.java:159]: if configurable then store the expiration TTL as value, and compare that later?
if value < clock.instant() then release and acquire

**Review by jlorper — APPROVED** [review]
A couple of suggestions, no blockers

File: java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java (hunk 2 above)
Comment (aperez-worklytics) [inline java/core/src/main/java/co/worklytics/psoxy/gateway/impl/oauth/OAuthRefreshTokenSourceAuthStrategy.java:225]: constant at class level?

File: java/impl/gcp/src/main/java/co/worklytics/psoxy/SecretManagerConfigService.java [hunk 6]
```diff
@@ -100,4 +112,80 @@ private String parameterName(ConfigProperty property) {
             return this.namespace + property.name();
         }
     }
+
+    static final java.time.Duration SECRET_TTL = java.time.Duration.ofDays(7);
```
Comment (aperez-worklytics) [inline java/impl/gcp/src/main/java/co/worklytics/psoxy/SecretManagerConfigService.java:116]: I'd move them at the top of the class

**Review by aperez-worklytics — APPROVED** [review]

File: java/impl/aws/src/main/java/co/worklytics/psoxy/aws/ParameterStoreConfigService.java (hunk 5 above)
Comment (eschultink) [reply java/impl/aws/src/main/java/co/worklytics/psoxy/aws/ParameterStoreConfigService.java:159]: that requires parsing the the value from `String` to `Instant`, vs just reading `lastModifiedDate` field directly from API

**Pull request merged**